From 4c87a92b0045c819a2a59d6e844d6f0d944107dc Mon Sep 17 00:00:00 2001
From: Olivier Moysan <olivier.moysan@foss.st.com>
Date: Wed, 26 Mar 2025 10:07:40 +0100
Subject: [PATCH] iio: adc: stm32-mdf: improve clock configuration check

As, the MDF shares its kernel clock with SAI, through the same flexgen,
the clocks settings may differ from initial probe configuration.

Rework stm32_mdf_adc_check_clock_config() function to check that all
clock constraints are still fulfilled at runtime.
If at least one condition is not met, try to restore the initial kernel
clock rate.

Also add traces related to clock configuration.

Change-Id: I770fc46fc14b5542dfcdc5604ed8ee071a80eea0
Signed-off-by: Olivier Moysan <olivier.moysan@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/445378
Domain-Review: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Reviewed-by: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 drivers/iio/adc/stm32-mdf-adc.c  | 90 +++++++++++++++++++-------------
 drivers/iio/adc/stm32-mdf-core.c | 10 ++--
 2 files changed, 61 insertions(+), 39 deletions(-)

diff --git a/drivers/iio/adc/stm32-mdf-adc.c b/drivers/iio/adc/stm32-mdf-adc.c
index fc7eb518034a..c5c0ae51c111 100644
--- a/drivers/iio/adc/stm32-mdf-adc.c
+++ b/drivers/iio/adc/stm32-mdf-adc.c
@@ -44,6 +44,10 @@
 #define MDF_IS_INTERLEAVED_FILT_NOT_0(adc)	({ typeof(adc) x = (adc);\
 						MDF_IS_INTERLEAVED_FILT(x) && !MDF_IS_FILTER0(x); })
 
+#define STM32_MDF_ERR_SCK_FREQ BIT(1)
+#define STM32_MDF_ERR_MODE_RATIO BIT(2)
+#define STM32_MDF_ERR_DECIM_RATIO BIT(3)
+
 struct stm32_mdf_dev_data {
 	int type;
 	int (*init)(struct device *dev, struct iio_dev *indio_dev);
@@ -698,28 +702,63 @@ static int stm32_mdf_adc_set_filters_config(struct iio_dev *indio_dev, unsigned
 	return stm32_mdf_adc_apply_filters_config(adc, scale);
 }
 
-static int stm32_mdf_adc_check_clock_config(struct stm32_mdf_adc *adc, unsigned long sck_freq)
+static int stm32_mdf_adc_check_clock_config(struct stm32_mdf_adc *adc, unsigned long *sck_freq)
 {
-	unsigned int ratio;
-	unsigned int decim_ratio;
+	unsigned long cck_expected_freq;
+	unsigned int ratio, cic_ratio;
+	int ret, error = 0;
 
-	ratio = DIV_ROUND_CLOSEST(adc->mdf->fproc, sck_freq);
-	decim_ratio = DIV_ROUND_CLOSEST(24, adc->decim_cic);
+	/*
+	 * The cck expected rate is set at probe from "clock-frequency" device tree property.
+	 * The serial interface clock sck (aka cck) is derived from MDF kernel clock frequency.
+	 * The kernel clock rate may have been changed by another consumer, or MDF dividers
+	 * may have been updated through clk_get_rate() call.
+	 * Ensure that kernel clock rate and MDF dividers, still provide the right sck/cck rate
+	 */
+	cck_expected_freq = stm32_mdf_core_get_cck(adc->mdf);
+	if (*sck_freq != cck_expected_freq) {
+		dev_dbg(adc->dev, "Wrong sck_freq [%lu]Hz. Expected [%lu]Hz.\n",
+			*sck_freq, cck_expected_freq);
+		error |= STM32_MDF_ERR_SCK_FREQ;
+	}
 
+	/*
+	 * Mode SPI:    check Fproc >= 4 * Fsck
+	 * Mode LF SPI: check Fproc >= 2 * Fsck
+	 */
+	ratio = DIV_ROUND_CLOSEST(adc->mdf->fproc, *sck_freq);
 	if ((adc->sitf->mode == STM32_MDF_MODE_SPI && ratio <= 4) ||
-	    (adc->sitf->mode == STM32_MDF_MODE_LF_SPI && ratio <= 2))
-		goto err;
+	    (adc->sitf->mode == STM32_MDF_MODE_LF_SPI && ratio <= 2)) {
+		dev_dbg(adc->dev, "Wrong Fproc/Fsck ratio [%d]. sitf mode [%d], RSFLT [%s]\n",
+			ratio, adc->sitf->mode, adc->rsflt_bypass ? "off" : "on");
+		error |= STM32_MDF_ERR_MODE_RATIO;
+	}
 
-	if (adc->rsflt_bypass && ratio <= decim_ratio)
-		goto err;
+	/* RSFLT enabled: check Fproc > 24 * Fsck / (MCICD+1) */
+	cic_ratio = DIV_ROUND_CLOSEST(24, adc->decim_cic);
+	if (!adc->rsflt_bypass && ratio <= cic_ratio)
+		error |= STM32_MDF_ERR_DECIM_RATIO;
+
+	/* If we found an error, try to restore clocks initial settings */
+	if (error) {
+		ret = stm32_mdf_core_restore_cck(adc->mdf);
+		if (ret < 0)
+			goto err;
+
+		*sck_freq = clk_get_rate(adc->sitf->sck);
+		if (!*sck_freq) {
+			ret = -EINVAL;
+			goto err;
+		}
+	}
 
 	return 0;
 
 err:
-	dev_err(adc->dev, "Wrong Fproc/Fsck ratio [%d] for sitf mode [%d] with RSFLT [%s]\n",
-		ratio, adc->sitf->mode, adc->rsflt_bypass ? "off" : "on");
+	dev_err(adc->dev, "Wrong clock configuration [0x%x]. Clock recovering returned [%d]\n",
+		error, ret);
 
-	return -EINVAL;
+	return ret;
 }
 
 static int mdf_adc_set_samp_freq(struct iio_dev *indio_dev, unsigned long sample_freq, int lock)
@@ -728,7 +767,6 @@ static int mdf_adc_set_samp_freq(struct iio_dev *indio_dev, unsigned long sample
 	struct device *dev = &indio_dev->dev;
 	unsigned int decim_ratio;
 	unsigned long delta, delta_ppm, sck_freq;
-	unsigned long cck_expected_freq;
 	int ret;
 
 	if (lock) {
@@ -753,25 +791,9 @@ static int mdf_adc_set_samp_freq(struct iio_dev *indio_dev, unsigned long sample
 	if (!sample_freq)
 		sample_freq = sck_freq / MDF_DEFAULT_DECIM_RATIO;
 
-	/*
-	 * MDF may share its parent clock with SAI, so kernel clock rate may have been changed.
-	 * The set_rate ops is called implicitly through clk_get_rate() call, and MDF dividers
-	 * may have been updated to keep the expected rate on cck clock. Check if sitf clock
-	 * frequency is still the expected one. If not, try to restore the kernel clock rate
-	 * for audio use case.
-	 */
-	cck_expected_freq = stm32_mdf_core_get_cck(adc->mdf);
-	if (sck_freq != cck_expected_freq) {
-		ret = stm32_mdf_core_restore_cck(adc->mdf);
-		if (ret < 0)
-			goto err;
-
-		sck_freq = clk_get_rate(adc->sitf->sck);
-		if (!sck_freq) {
-			ret = -EINVAL;
-			goto err;
-		}
-	}
+	ret = stm32_mdf_adc_check_clock_config(adc, &sck_freq);
+	if (ret)
+		goto err;
 
 	decim_ratio = DIV_ROUND_CLOSEST(sck_freq, sample_freq);
 
@@ -803,10 +825,6 @@ static int mdf_adc_set_samp_freq(struct iio_dev *indio_dev, unsigned long sample
 	if (ret < 0)
 		goto err;
 
-	ret = stm32_mdf_adc_check_clock_config(adc, sck_freq);
-	if (ret < 0)
-		goto err;
-
 	adc->sample_freq = DIV_ROUND_CLOSEST(sck_freq, decim_ratio);
 	adc->decim_ratio = decim_ratio;
 
diff --git a/drivers/iio/adc/stm32-mdf-core.c b/drivers/iio/adc/stm32-mdf-core.c
index 23360563e222..6071638b6569 100644
--- a/drivers/iio/adc/stm32-mdf-core.c
+++ b/drivers/iio/adc/stm32-mdf-core.c
@@ -194,13 +194,16 @@ EXPORT_SYMBOL_GPL(stm32_mdf_core_get_cck);
 int stm32_mdf_core_lock_kclk_rate(struct stm32_mdf *mdf)
 {
 	struct stm32_mdf_priv *priv = to_stm32_mdf_priv(mdf);
+	struct device *dev = &priv->pdev->dev;
 	int ret;
 
 	if (!atomic_read(&priv->n_active_ch)) {
 		/* Request rate exclusivity on kernel clock right now */
 		ret = clk_rate_exclusive_get(priv->kclk);
-		if (ret < 0)
+		if (ret < 0) {
+			dev_err(dev, "Failed to get exclusive rate on kernel clock: [%d]\n", ret);
 			return ret;
+		}
 
 		priv->exclusive_flg = true;
 	}
@@ -228,11 +231,12 @@ int stm32_mdf_core_restore_cck(struct stm32_mdf *mdf)
 
 	ret = clk_set_rate(priv->kclk, priv->kclk_rate);
 	if (ret) {
-		dev_err(dev, "Changing MDF kernel clock rate returned error %d\n", ret);
+		dev_err(dev, "Failed to change kernel clock rate from [%lu]Hz to [%lu]Hz: [%d]\n",
+			clk_get_rate(priv->kclk), priv->kclk_rate, ret);
 		return ret;
 	}
 
-	dev_dbg(dev, "MDF kernel clock rate changed to [%lu]Kz.\n", priv->kclk_rate);
+	dev_dbg(dev, "MDF kernel clock rate changed to [%lu]Hz.\n", priv->kclk_rate);
 
 	return 0;
 }
-- 
2.50.1

